crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -1413,6 +1413,131 @@ impl<TX: DbTx> BlockReader for DatabaseProvider<TX> {
+        }
+
+        let len = range.end().saturating_sub(*range.start()) as usize;
+        let mut blocks = Vec::with_capacity(len);
+
+        let mut headers_cursor = self.tx.cursor_read::<tables::Headers>()?;
+        let mut ommers_cursor = self.tx.cursor_read::<tables::BlockOmmers>()?;
+        let mut withdrawals_cursor = self.tx.cursor_read::<tables::BlockWithdrawals>()?;
+        let mut block_body_cursor = self.tx.cursor_read::<tables::BlockBodyIndices>()?;
+        let mut tx_cursor = self.tx.cursor_read::<tables::Transactions>()?;
+        let mut senders_cursor = self.tx.cursor_read::<tables::TxSenders>()?;
+        for num in range {
+            if let Some((_, header)) = headers_cursor.seek_exact(num)? {
+                // If the body indices are not found, this means that the transactions either do
+                // not exist in the database yet, or they do exit but are
+                // not indexed. If they exist but are not indexed, we don't
+                // have enough information to return the block anyways, so
+                // we skip the block.
+                if let Some((_, block_body_indices)) = block_body_cursor.seek_exact(num)? {
+                    let tx_range = block_body_indices.tx_num_range();
+                    let body = if tx_range.is_empty() {
+                        Vec::new()
+                    } else {
+                        tx_cursor
+                            .walk_range(tx_range.clone())?
+                            .map(|result| result.map(|(tx_number, tx)| (tx_number, tx.with_hash())))
+                            .collect::<Result<Vec<_>, _>>()?
+                    };
+
+                    // If we are past shanghai, then all blocks should have a withdrawal list,
+                    // even if empty
+                    let withdrawals =
+                        if self.chain_spec.is_shanghai_active_at_timestamp(header.timestamp) {
+                            Some(
+                                withdrawals_cursor
+                                    .seek_exact(num)?
+                                    .map(|(_, w)| w.withdrawals)
+                                    .unwrap_or_default(),
+                            )
+                        } else {
+                            None
+                        };
+                    let ommers = if self.chain_spec.final_paris_total_difficulty(num).is_some() {
+                        Vec::new()
+                    } else {
+                        ommers_cursor.seek_exact(num)?.map(|(_, o)| o.ommers).unwrap_or_default()
+                    };
+
+                    // Walking range to get the TxNumber along with the Address.
+                    let mut senders =
+                        senders_cursor.walk_range(tx_range)?.collect::<Result<Vec<_>, _>>()?;
+
+                    // This approach is similar to what is done in
+                    // `get_take_block_transaction_range`.
```

Comment:
can we use the existing function here?

Reply:
`block_range` can be used here but `get_take_block_transaction_range` cannot be used as it will require changing the trait bounds of the `BlockReader` implementation to add `DbTxMut`.

Reply:
Moreover, since this is a read-only method and `TAKE` would always be false in `get_take_block_transaction_range`. I don't think it makes sense to use it.